In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import scipy as sp
import statsmodels.api as sm
import seaborn as sns
import bambi as bmb
import arviz as az

def remove_nulls_simple(data_subset, variables):
        # Drop rows where any of the specified variable columns have nulls
        data_subset_cleaned = data_subset.dropna(subset=variables, ignore_index=True)
        # Subsets the cleaned dataframe back into only chosen variables
        vars_cleaned = data_subset_cleaned[variables]
        return vars_cleaned

def list_to_str(list_name):
    string = " + ".join(list_name)
    return string

pa_gs = pd.read_csv('data/pa_data.csv')
temps = pd.read_csv("data/pa_temps_supplement.csv")
#combining main csv dataset with supplementary temperature data csv
pa_gs = pa_gs.merge(temps, on=["station_id", "year"], how="left")

#limiting dataset to just time interval of interest
pa_gs = pa_gs[pa_gs['year'] >= 1960]
#dropping non-numeric columns
pa_gs = pa_gs.drop(['last_spring_frost_date', 'first_fall_frost_date', 'station_name','state'], axis=1)

In [3]:
stn_rdm_effect_model_formula = bmb.Formula("growing_season_length ~ year + (1|station_id)", "sigma ~ year + (1|station_id)")
stn_rdm_effect_model = bmb.Model(stn_rdm_effect_model_formula, pa_gs, dropna = True)
stn_rdm_effect_output = stn_rdm_effect_model.fit()

                                                            Grad                                                  
  Progress               Draw        Divergen…   Step size   evals       Speed                Elapsed    Remaini…  
 ───────────────────────────────────────────────────────────────────────────────────────────────────────────────── 
  ━━━━━━━━━━━━━━━━━━━━   2000        0           0.168       15          162.43 draws/s       0:00:12    0:00:00   
  ━━━━━━━━━━━━━━━━━━━━   2000        0           0.219       31          160.89 draws/s       0:00:12    0:00:00   
  ━━━━━━━━━━━━━━━━━━━━   2000        0           0.190       31          125.66 draws/s       0:00:15    0:00:00   
  ━━━━━━━━━━━━━━━━━━━━   2000        0           0.182       31          127.10 draws/s       0:00:15    0:00:00

Sampling 4 chains for 1_000 tune and 1_000 draw iterations (4_000 + 4_000 draws total) took 18 seconds.
The rhat statistic is larger than 1.01 for some parameters. This indicates problems during sampling. See https://arxiv.org/abs/1903.08008 for details
The effective sample size per chain is smaller than 100 for some parameters.  A higher number is needed for reliable rhat and ess computation. See https://arxiv.org/abs/1903.08008 for details


In [6]:
az.summary(stn_rdm_effect_output).head(6)

,mean,sd,eti89_lb,eti89_ub,ess_bulk,ess_tail,r_hat,mcse_mean,mcse_sd
Intercept,-609.7,28.7,-660,-560,4587,3238,1.00,0.42,0.3
year,0.3892,0.0144,0.37,0.41,4580,3327,1.00,0.00021,0.00015
1|station_id_sigma,24.4,1.32,22,27,414,492,1.01,0.067,0.049
sigma_Intercept,10.58,1.11,8.8,12,5026,3050,1.00,0.016,0.011
sigma_year,-0.00387,0.00056,-0.0048,-0.003,5008,3015,1.00,7.9e-06,5.5e-06
sigma_1|station_id_sigma,0.1595,0.0143,0.14,0.18,1514,2111,1.00,0.00037,0.00027


#### Analysis of Station Random Effect Model
* Station Random Effect Mean of Response is 24.4
* Station Random Effect Variance of Response is 17.292%
    * would expect a typical spread of +/- 17% in the variance for stations

In [7]:
cov_all = list_to_str(['dtr_annual','tmean_spring','tmean_fall','latitude','longitude','tmax_annual','oni_annual',
                'nao_annual','pna_annual','amo_annual','sst_north_atlantic','pwat_station','dewpoint_station',
                'soil_moisture_station','cloud_cover_station','evaporation_station', 'dtr_spring',
                'sst_gulf_mexico','pwat_southeast_us','dewpoint_2m_southeast_us','soil_moisture_southeast_us',
                'cloud_cover_southeast_us','evaporation_southeast_us'])

formula_no_sigma = bmb.Formula(f"growing_season_length ~ {cov_all} + year")
model_no_sigma = bmb.Model(formula_no_sigma, data=pa_gs, dropna = True)
output_no_sigma = model_no_sigma.fit()

                                                            Grad                                                  
  Progress               Draw        Divergen…   Step size   evals       Speed                Elapsed    Remaini…  
 ───────────────────────────────────────────────────────────────────────────────────────────────────────────────── 
  ━━━━━━━━━━━━━━━━━━━━   2000        0           0.128       31          325.37 draws/s       0:00:06    0:00:00   
  ━━━━━━━━━━━━━━━━━━━━   2000        0           0.259       31          318.00 draws/s       0:00:06    0:00:00   
  ━━━━━━━━━━━━━━━━━━━━   2000        0           0.235       31          310.97 draws/s       0:00:06    0:00:00   
  ━━━━━━━━━━━━━━━━━━━━   2000        0           0.194       31          320.68 draws/s       0:00:06    0:00:00

Sampling 4 chains for 1_000 tune and 1_000 draw iterations (4_000 + 4_000 draws total) took 10 seconds.


In [8]:
az.summary(output_no_sigma)

,mean,sd,eti89_lb,eti89_ub,ess_bulk,ess_tail,r_hat,mcse_mean,mcse_sd
sigma,17.365,0.153,17,18,6197,3309,1.00,0.0019,0.0014
Intercept,52,69,-58,160,3082,3166,1.00,1.2,0.86
dtr_annual,-10.12,0.44,-11,-9.4,3020,2820,1.00,0.008,0.0058
tmean_spring,1.157,0.268,0.73,1.6,3538,3022,1.00,0.0045,0.0033
tmean_fall,5.087,0.266,4.7,5.5,4090,3325,1.00,0.0042,0.0029
latitude,-2.17,0.59,-3.1,-1.2,4138,3124,1.00,0.0091,0.0068
longitude,1.437,0.21,1.1,1.8,3448,2872,1.00,0.0036,0.0026
tmax_annual,4.51,0.45,3.8,5.2,2951,2892,1.00,0.0083,0.0058
oni_annual,0.02,0.492,-0.78,0.78,5360,2957,1.00,0.0067,0.0047
nao_annual,-0.41,0.83,-1.7,0.89,4004,3252,1.00,0.013,0.0094


In [3]:
stn_time_rdm_effect_model_formula = bmb.Formula("growing_season_length ~ year + (year|station_id)", "sigma ~ year + (year|station_id)")
stn_time_rdm_effect_model = bmb.Model(stn_time_rdm_effect_model_formula, pa_gs, dropna = True)
stn_time_rdm_effect_output = stn_time_rdm_effect_model.fit()

Initializing NUTS using jitter+adapt_diag...


SamplingError: Initial evaluation of model at starting point failed!
Starting values:
{'Intercept': array(163.46862303), 'year': array(-0.18768903), '1|station_id_sigma_log__': array(9.63706994), '1|station_id_offset': array([ 0.1743314 , -0.6684252 , -0.85332349, -0.50430432,  0.25584946,
        0.426941  , -0.07965472, -0.74567353,  0.89186991, -0.86166394,
       -0.92531301, -0.20746083,  0.32404097, -0.15325118,  0.0286636 ,
       -0.86505139, -0.6202056 , -0.97583564, -0.08690775, -0.65638616,
        0.20512408, -0.7627213 ,  0.05667383, -0.31105819,  0.30009168,
        0.91907051,  0.79718454,  0.71434897, -0.48241094, -0.424065  ,
        0.29497302, -0.79546633, -0.56923981,  0.43142262,  0.14379529,
        0.48538454, -0.01226498,  0.71116248,  0.76301347,  0.2895076 ,
       -0.27272701,  0.20749776, -0.551816  , -0.20772299,  0.15961143,
       -0.87355978,  0.73981303,  0.88810188,  0.51746955,  0.00108635,
       -0.49323454, -0.6117877 , -0.10404278, -0.91004631,  0.5627829 ,
       -0.43897761, -0.19927944, -0.31197726,  0.28493689,  0.27158173,
        0.91312855, -0.15830846,  0.91904737,  0.20208485, -0.90179015,
       -0.06132797,  0.10321077, -0.67240353,  0.52838059,  0.26028427,
       -0.25615148, -0.3633942 , -0.39945233, -0.15283372,  0.05372127,
        0.48374468, -0.48866371, -0.83400768,  0.82688916, -0.47958167,
       -0.40633831, -0.79202056, -0.90286023, -0.29067306,  0.4667589 ,
        0.97942656, -0.85896669,  0.28029211, -0.9655389 , -0.27954686,
       -0.30125517, -0.20143714, -0.56411279, -0.80775868,  0.81671989,
        0.71644924, -0.89260849,  0.27181757,  0.7055729 ,  0.77382353,
       -0.36316273, -0.8105204 ,  0.13695493,  0.03238387,  0.42171989,
        0.80663383, -0.91876977, -0.44452249,  0.777549  , -0.00289787,
       -0.5659958 ,  0.57820301, -0.25995764,  0.12206077, -0.69034775,
        0.24622285, -0.8084009 ,  0.33054539,  0.74302673, -0.12852325,
        0.30630491, -0.56916675, -0.13194491,  0.89397481,  0.63399787,
        0.3162676 ,  0.86156721, -0.83404514, -0.51172298, -0.18938378,
       -0.27350164, -0.17765194,  0.35787507,  0.17020497,  0.42271172,
        0.40007099,  0.61540614, -0.06995295,  0.26700667, -0.47507356,
        0.44329022,  0.42408934,  0.64636171,  0.10955437,  0.37135056,
       -0.03000655, -0.82317867,  0.63681372, -0.01778061, -0.65864612,
       -0.83897946,  0.27656897, -0.64160963, -0.45079722, -0.55773705,
       -0.47549065,  0.88657894, -0.416022  , -0.49700254, -0.61222763,
       -0.99085743,  0.26674198,  0.10932729,  0.52121086,  0.10874096,
       -0.45979486, -0.47994547,  0.79802532,  0.36669302,  0.27957743,
        0.31378967, -0.6529863 ,  0.6179156 ,  0.76116745,  0.34455744,
       -0.05523871,  0.16795336, -0.63952208,  0.95733426,  0.65949427]), 'year|station_id_sigma_log__': array(1.08630799), 'year|station_id_offset': array([-0.81721168,  0.83993409,  0.76874882, -0.9724573 , -0.60870158,
       -0.82587758, -0.16926621, -0.67861451, -0.97403257,  0.32025302,
        0.78387464, -0.12002817,  0.38472856,  0.98312112,  0.79398743,
        0.78365089, -0.62194126,  0.13840633, -0.95483503,  0.2881576 ,
       -0.65230301,  0.01885192,  0.12141946, -0.5300254 ,  0.03674354,
       -0.20044896,  0.63859897,  0.40165096, -0.5431331 ,  0.36030844,
       -0.30130271, -0.21358253, -0.93287049,  0.77564616, -0.59851532,
        0.2339541 , -0.51672169,  0.6844277 , -0.22055224,  0.27738664,
       -0.18384709,  0.07053807, -0.26058461,  0.31287278, -0.48441815,
        0.68766997,  0.40702905, -0.17777226, -0.9312388 , -0.25914182,
       -0.90324026, -0.99791321, -0.41367465, -0.58957547, -0.90903158,
        0.9649173 ,  0.57506585,  0.05407837,  0.24600477,  0.9914984 ,
        0.10032733,  0.53170251, -0.9826075 ,  0.81849322,  0.50299839,
       -0.17667139, -0.61415248,  0.42541773, -0.39489781,  0.00726267,
       -0.17697817, -0.89761843, -0.64009477, -0.7902814 , -0.041342  ,
        0.47589409,  0.06543938, -0.06854823,  0.05236487, -0.78991018,
       -0.52318384, -0.20654301,  0.66315025, -0.4440127 ,  0.9244196 ,
        0.66235638, -0.50141835,  0.3253234 ,  0.19077306, -0.52830023,
       -0.88637247, -0.80598177, -0.64129538,  0.9486184 ,  0.33902442,
        0.19902599,  0.82913245,  0.53475297,  0.2665917 ,  0.19841454,
       -0.61975961, -0.1041584 ,  0.23871161,  0.88809481, -0.78537917,
        0.61362321,  0.9865828 , -0.4888515 , -0.83951358,  0.90623443,
        0.55896121, -0.0198786 , -0.72078237, -0.88516393,  0.87233645,
        0.88484255,  0.57061963,  0.27369226,  0.55503782,  0.63220969,
       -0.47355737,  0.68545526,  0.32143693, -0.22658051, -0.16264799,
        0.54608156,  0.1290473 , -0.4192658 , -0.86146093,  0.43845053,
       -0.64211278,  0.15789546,  0.37860936, -0.05332642, -0.48757078,
       -0.53586148, -0.07118754,  0.68102006,  0.98750924, -0.28039194,
        0.18512019, -0.62296725,  0.45314502,  0.68032493,  0.90237563,
        0.6641826 , -0.50038746,  0.64856961,  0.47121648,  0.41279843,
        0.93992718, -0.646358  , -0.00629539, -0.83102638,  0.04111134,
        0.71136207,  0.2298261 ,  0.29437359, -0.81494074,  0.35746747,
       -0.86082817, -0.25538604, -0.76194978, -0.0962018 , -0.79098311,
       -0.11096496,  0.76312069,  0.10737633, -0.42913359, -0.88707038,
       -0.7594082 , -0.8245966 , -0.8018776 ,  0.73061269,  0.10850775,
        0.50453515,  0.23365249,  0.20773823,  0.16011151, -0.82462981]), 'sigma_Intercept': array(-0.21586675), 'sigma_year': array(0.46752592), 'sigma_1|station_id_sigma_log__': array(0.02279795), 'sigma_1|station_id_offset': array([-0.13162026,  0.37150718, -0.06725846,  0.46842194,  0.65991235,
        0.20615456, -0.72276139,  0.1286135 , -0.78345643, -0.60446194,
        0.38619232,  0.72531404, -0.82028016, -0.40644131, -0.47131241,
        0.24999544,  0.98562757,  0.78279912, -0.88463435, -0.67091711,
       -0.0432213 ,  0.64338761, -0.73062753,  0.71338483, -0.99110229,
        0.73345429,  0.12255499,  0.15245897,  0.94253806,  0.4166668 ,
       -0.61046436,  0.24527343,  0.96721763,  0.75915543,  0.31496868,
       -0.28263019, -0.18260134, -0.95726623, -0.27368789,  0.70793943,
       -0.16317742, -0.72984315, -0.83653793,  0.75436745,  0.23820116,
       -0.49669496,  0.74410248,  0.37779427, -0.3643516 , -0.29551855,
        0.28525473,  0.76685619, -0.8835594 ,  0.71269009, -0.64323096,
        0.40125489,  0.6104274 ,  0.57712474,  0.01184473,  0.55328514,
        0.20003504, -0.53865641,  0.37892805, -0.06993565,  0.67241881,
       -0.13921345,  0.42567309, -0.58380156,  0.65463996, -0.63688608,
        0.63018087,  0.18184141,  0.75453458,  0.31954234, -0.74386501,
       -0.57738242,  0.54239874,  0.20561542,  0.98720611, -0.73753276,
       -0.4948125 ,  0.1942854 ,  0.29837125, -0.90974453,  0.19612532,
       -0.54179297, -0.96886071, -0.01148151, -0.27408007, -0.2422585 ,
       -0.242035  ,  0.47808393, -0.18572601,  0.44350098,  0.05872834,
       -0.5785348 ,  0.35921546,  0.019388  , -0.68297765, -0.11440014,
       -0.07225623, -0.46721303,  0.11867459,  0.19464513, -0.84415932,
        0.20799951,  0.00602358,  0.13840155,  0.25456068, -0.74824092,
       -0.88103372,  0.73764898, -0.115203  , -0.83276515,  0.54338995,
       -0.53614901,  0.1508898 ,  0.66213119, -0.42064368, -0.78726918,
        0.52634986, -0.38168319, -0.83331288,  0.83460415, -0.1967972 ,
        0.73232272, -0.96126406,  0.20785771, -0.12005489, -0.76682866,
        0.31446669,  0.48756998,  0.84309219, -0.08763305, -0.36944064,
       -0.96885843,  0.08239067,  0.33649408,  0.86131483, -0.38029056,
        0.02876625, -0.5650131 ,  0.83289895, -0.09552779,  0.10389338,
        0.92830988, -0.28837133, -0.90801312, -0.87431327, -0.42192256,
       -0.27529408,  0.15467397, -0.781005  ,  0.47276647,  0.74150405,
       -0.75026478,  0.58112768,  0.63711005,  0.77263462,  0.03902683,
        0.68475172,  0.71049569,  0.728337  ,  0.35072825,  0.480503  ,
        0.82736096, -0.1068741 ,  0.13686423, -0.00941884, -0.79698948,
       -0.71903085,  0.51634189, -0.20906027, -0.44493848,  0.04067645,
        0.10698457,  0.06055596, -0.78389269, -0.77215078, -0.02877993]), 'sigma_year|station_id_sigma_log__': array(0.34213374), 'sigma_year|station_id_offset': array([ 0.41587246,  0.47573489,  0.16643407, -0.59739146,  0.90198891,
       -0.63648109,  0.22457083,  0.46840244, -0.18479902,  0.42273551,
       -0.45559551,  0.16773979,  0.60672725,  0.69758544,  0.37407168,
       -0.80780089,  0.96174058, -0.8990859 ,  0.68232272,  0.41888295,
        0.48264463, -0.09069428, -0.1763349 , -0.21481629, -0.99551385,
        0.53999336, -0.20004162,  0.94716246, -0.74971158, -0.74662253,
        0.03908601, -0.32167694, -0.64216035, -0.4937215 ,  0.23944828,
        0.39474624,  0.70833353,  0.52906105, -0.05352415,  0.44281111,
       -0.55457395, -0.05627053,  0.90929858, -0.41647865,  0.27194211,
        0.31276134, -0.19988705,  0.64493125, -0.78400256, -0.79042239,
       -0.3319393 , -0.72322966,  0.95227856, -0.20182544, -0.85455281,
        0.27351032, -0.03833835,  0.79347954, -0.96720563,  0.1682907 ,
       -0.50264204,  0.26551256,  0.42202463,  0.28921449, -0.70224458,
        0.19608103, -0.15212495, -0.3141455 , -0.97057812,  0.4226879 ,
        0.66202889,  0.77023328, -0.70607021, -0.01244137,  0.8136408 ,
        0.3043342 , -0.80916119, -0.86661061,  0.92436535, -0.12464675,
        0.06532814,  0.45827211,  0.94396432,  0.82501265, -0.08300811,
       -0.08022916,  0.02688677, -0.92889968, -0.48324026,  0.4132314 ,
       -0.55704445,  0.582354  , -0.44203896,  0.65773522,  0.49007778,
        0.48881686, -0.03594824,  0.01231697, -0.54716102, -0.60568521,
        0.1679231 , -0.9397622 ,  0.61557765,  0.10320424,  0.46376519,
        0.24511371,  0.4454602 , -0.1160605 ,  0.19798106, -0.8476364 ,
        0.75606375, -0.56300727,  0.10692224, -0.77101469,  0.4534866 ,
       -0.28946946,  0.37810269, -0.78384493, -0.86128509, -0.19048895,
       -0.22457362, -0.08979577, -0.94732299, -0.2099714 , -0.61852385,
       -0.16928977,  0.99581023, -0.54268307, -0.90843996,  0.26241025,
       -0.74100596,  0.06387461, -0.12265839, -0.92036882, -0.31546816,
        0.8111261 , -0.93499657,  0.16766369,  0.1268752 , -0.77316128,
       -0.53966638,  0.40256963, -0.90480149,  0.50152792, -0.00136074,
       -0.691122  , -0.84744264,  0.98192727, -0.01307915, -0.4899521 ,
       -0.97088879, -0.93182944,  0.37932751,  0.7351446 , -0.75151835,
        0.0774872 ,  0.35935809,  0.47578729,  0.06162906,  0.97075862,
       -0.48092214,  0.82182824,  0.30083285, -0.43917542, -0.98334176,
       -0.77336596,  0.87380454, -0.32280792, -0.58408859, -0.46009354,
        0.24675108,  0.9075962 , -0.67836998,  0.96781294,  0.234429  ,
        0.83761746,  0.07949096,  0.85113639,  0.36551926, -0.8684046 ])}

Logp initial evaluation results:
{'Intercept': np.float64(-9.93), 'year': np.float64(-2.33), '1|station_id_sigma': np.float64(-1.36), '1|station_id_offset': np.float64(-193.3), 'year|station_id_sigma': np.float64(-0.81), 'year|station_id_offset': np.float64(-196.77), 'sigma_Intercept': np.float64(-0.94), 'sigma_year': np.float64(-1.03), 'sigma_1|station_id_sigma': np.float64(-0.73), 'sigma_1|station_id_offset': np.float64(-193.79), 'sigma_year|station_id_sigma': np.float64(-0.87), 'sigma_year|station_id_offset': np.float64(-196.51), 'growing_season_length': np.float64(-inf)}
You can call `model.debug()` for more details.